pkg/tfbridge/x/tokens.go

```diff
@@ -649,62 +649,96 @@ func aliasResource(
 	if res == nil {
 		return
 	}
-
-	safeRange := func(r shim.Resource, f func(string, shim.Schema)) {
+	var walk func(*fieldHistory, *b.SchemaInfo, shim.Schema) (bool, bool)
+	walkResource := func(r shim.Resource, hist *map[string]*fieldHistory, info *map[string]*b.SchemaInfo) (bool, bool) {
 		if r == nil {
-			return
+			return hist != nil, info != nil
 		}
 		m := r.Schema()
 		if m == nil {
-			return
+			return hist != nil, info != nil
 		}
+
+		var rHasH, rHasI bool
+
 		m.Range(func(k string, v shim.Schema) bool {
-			f(k, v)
+			h, hasH := getNonNil(hist, k)
+			i, hasI := getNonNil(info, k)
+			keepH, keepI := walk(h, i, v)
+
+			hasH = hasH || keepH
+			hasI = hasI || keepI
+
+			if !hasH {
+				delete(*hist, k)
+			}
+			if !hasI {
+				delete(*info, k)
+			}
+
+			rHasH = rHasH || hasH
+			rHasI = rHasI || hasI
+
 			return true
 		})
+
+		return rHasH, rHasI
 	}
 
-	var walk func(*fieldHistory, *b.SchemaInfo, shim.Schema)
-	walk = func(h *fieldHistory, info *b.SchemaInfo, schema shim.Schema) {
+	walk = func(h *fieldHistory, info *b.SchemaInfo, schema shim.Schema) (hasH bool, hasI bool) {
```

Comment:
I think there is a lot of state context here that gets confusing because the variable names are so similar.
For example, we return `hasH` and `hasI` here but on line 667 we assign them to `keepH` and `keepI` and it would be great to know what is prompting us to "keep" info or history.

Reply:
I added that comment.